src/blockMonitor/blockchainMachine.ts

```diff
@@ -1,26 +1,64 @@
 import { BlockProcessor } from "./blockProcessor";
 import { IBlockStub, StartStopService, ApplicationError } from "../dataEntities";
 import { Component, AnchorState, ComponentAction } from "./component";
+import { ActionStore, ActionAndId } from "./actionStore";
+import { BlockItemStore } from "../dataEntities/block";
+import { Lock } from "../utils/lock";
 
-interface ComponentAndStates {
-    component: Component<AnchorState, IBlockStub, ComponentAction>;
-    states: WeakMap<IBlockStub, AnchorState>;
-}
-
-// Generic class to handle the anchor statee of a blockchain state machine
+/**
+ * Generic class to handle the anchor state of a blockchain state machine.
+ *
+ * For each block that is added to the cache (and for each added component), this will compute the new anchor state for
+ * that component. Moreover, every time a "new head" event is emitted, this class will use the component's `detectChanges`
+ * function to compute the appropriate actions, by comparing the newly computed anchor state with the anchor state of the
+ * closest ancestor that was emitted as "new head". Since the latter might no longer be in the BlockCache, its anchor state
+ * is propagated in each subsequent block; thus, every block stores its anchor state, and the "closest emitted ancestor"'s one.
+ */
 export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopService {
-    private componentsAndStates: ComponentAndStates[] = [];
+    private components: Component<AnchorState, IBlockStub, ComponentAction>[] = [];
+    private componentNames: Set<string> = new Set();
+
+    // lock used to make sure that all events are processed in order
+    private lock = new Lock();
+
+    /**
+     * Runs all the actions in `actionAndIds` for `component`. Actions are all executed in parallel, and each action is removed
+     * from the ActionStore upon completion.s
+     */
+    private runActionsForComponent(component: Component<AnchorState, IBlockStub, ComponentAction>, actionAndIds: Iterable<ActionAndId>) {
+        // Side effects must be thread safe, so we can execute them concurrently
+        // Note that actions are executed in background and not awaited for in here.
+        [...actionAndIds].forEach(async a => {
+            try {
+                await component.applyAction(a.action);
+                this.actionStore.removeAction(component.name, a);
```

Comment:
Shouldnt we be awaiting the removeAction?

Reply:
Ooops, funny I added the try/catch exactly to log errors thrown from the removeAction, but I forgot to await it. >:)
Fixed in 2f123df.